yb-voyager/src/query/queryissue/detectors_test.go

```diff
@@ -530,37 +542,45 @@ RETURNING id,
                      xmlattributes(id AS "ID"),
                      xmlforest(name AS "Name", salary AS "NewSalary", xmin AS "TransactionStartID", xmax AS "TransactionEndID"));`,
 	}
-	expectedConstructs := []string{ADVISORY_LOCKS_NAME, SYSTEM_COLUMNS_NAME, XML_FUNCTIONS_NAME}
+	expectedIssueTypes := mapset.NewThreadUnsafeSet[string]([]string{ADVISORY_LOCKS, SYSTEM_COLUMNS, XML_FUNCTIONS}...)
 
-	detectors := []UnsupportedConstructDetector{
-		NewFuncCallDetector(),
-		NewColumnRefDetector(),
-		NewXmlExprDetector(),
-	}
 	for _, sql := range combinationSqls {
+		detectors := []UnsupportedConstructDetector{
+			NewFuncCallDetector(sql),
+			NewColumnRefDetector(sql),
+			NewXmlExprDetector(sql),
+		}
 		parseResult, err := queryparser.Parse(sql)
 		assert.NoError(t, err)
 
 		visited := make(map[protoreflect.Message]bool)
-		unsupportedConstructs := []string{}
 
 		processor := func(msg protoreflect.Message) error {
 			for _, detector := range detectors {
 				log.Debugf("running detector %T", detector)
-				constructs, err := detector.Detect(msg)
+				err := detector.Detect(msg)
 				if err != nil {
 					log.Debugf("error in detector %T: %v", detector, err)
 					return fmt.Errorf("error in detectors %T: %w", detector, err)
 				}
-				unsupportedConstructs = lo.Union(unsupportedConstructs, constructs)
 			}
 			return nil
 		}
 
 		parseTreeMsg := queryparser.GetProtoMessageFromParseTree(parseResult)
 		err = queryparser.TraverseParseTree(parseTreeMsg, visited, processor)
 		assert.NoError(t, err)
-		assert.ElementsMatch(t, expectedConstructs, unsupportedConstructs, "Detected constructs do not exactly match the expected constructs. Expected: %v, Actual: %v", expectedConstructs, unsupportedConstructs)
+
+		var allIssues []QueryIssue
+		for _, detector := range detectors {
+			allIssues = append(allIssues, detector.GetIssues()...)
+		}
+		issueTypesDetected := mapset.NewThreadUnsafeSet[string]()
+		for _, issue := range allIssues {
+			issueTypesDetected.Add(issue.Type)
+		}
+
+		assert.True(t, expectedIssueTypes.Equal(issueTypesDetected), "Expected issue types do not match the detected issue types. Expected: %v, Actual: %v", expectedIssueTypes, issueTypesDetected)
```

Comment:
Can we enhance this assert to count of issues for each issue Type?

Reply:
So, the thing is: 
Ideally, there should only be 1 issue of each type.
The code to return a single issue for XML functiosn is in parserIssueDetector. So instead of checking the counts here, i added a separate test case `TestSingleXMLIssueIsDetected` to ensure that only one is finally returned. 

Moving forward, I think we should majorly focus on writing tests at the parserIssueDetector.GetIssues() rather than at the detector directly (unless ofcourse there is some corner case we want to specifically test).